File: 290-WordPattern/word_pattern.rs
fn word_pattern(pt: String, s: String) -> bool {
    if pt.is_empty() && s.is_empty() {
        true
    } else if pt == "" || s == "" {
        false
    } else {
        let p = pt.chars().next().unwrap();
        let s_words: &Vec<&str> = &s.split_ascii_whitespace().collect();
        let s_first = s_words[0];
        pt.chars()
            .map(|c| c == p)
            .eq(s_words.into_iter().map(|word| word == &s_first))
            && word_pattern(
                pt.chars().filter(|c| c != &p).collect(),
                s_words // This is a mess.
                    .to_owned()
                    .into_iter()
                    .filter(|w| w != &s_first)
                    .collect::<Vec<&str>>()
                    .join(" "),
            )
    }
}

#[cfg(test)]
mod tests {
    use crate::word_pattern;
    #[test]
    fn ex1() {
        assert!(word_pattern("abba".to_string(), "dog cat cat dog".to_string()))
    }
    #[test]
    fn ex2() {
        assert!(!word_pattern("abba".to_string(), "dog cat cat fish".to_string()))
    }
    #[test]
    fn ex3() {
        assert!(!word_pattern("aaaa".to_string(), "dog cat cat dog".to_string()))
    }
}

File: 290-WordPattern/wordPattern.js
"use strict";
const compareArrays = (a, b) => a.length === b.length && a.every((element, index) => element === b[index]);
const wordPattern = (pattern, w) => {
    if (pattern == "" && w == "") {
        return true;
    }
    else if (pattern == "" || w == "") {
        return false;
    }
    const p = pattern.charAt(0);
    const sWords = w.split(" ");
    //console.log(sWords, pattern);
    const s = sWords[0];
    //console.log(p, s);
    return compareArrays((pattern.split("").map(c => c === p)), (sWords.map(w => w === s))) &&
        wordPattern((pattern.replaceAll(p, "")), (sWords.filter(w => w !== s)).join(" "));
};
console.log(wordPattern("abba", "dog cat cat dog"));
console.log(wordPattern("abba", "dog cat cat fish"));
console.log(wordPattern("aaaa", "dog cat cat dog"));
